scalding-core/src/main/scala/com/twitter/scalding/RichPipe.scala

```diff
@@ -157,6 +175,59 @@ object RichPipe extends java.io.Serializable {
         false
     }
   }
+
+  /**
+    * Special handling for cases where one side of the hashjoin is merged
+    *  with the hashjoin result. Cascading no longer allows it (as of 3.0),
+    *  so we insert checkpoints and/or intermediate merge stages as appropriate
+    *
+    * @param head the first pipe to be merged
+    * @param tail a list of other pipes to be merged within the first
+    *
+    * @return an updated list of pipes, which can safely be merged
+    */
+  private[scalding] def mergeAvoidingHashes(head: Pipe, tail: List[Pipe]): Pipe = {
+
+    // we make use of the fact that the pipe merge operation is not just associative but also commutative
+    val pipes = head :: tail
+    val (colliding, uncolliding) = pipes.partition(p => pipes.exists(o => (o != p) && isHashJoinedWithPipe(p, o)))
+
+    val (innerColliding, innerUncolliding) = colliding.partition(p =>
+      colliding.exists(o => (o != p) && (isHashJoinedWithPipe(p, o) || isHashJoinedWithPipe(o, p))))
+    /* innerUncolliding pipes collide with some pipes in the uncolliding set, but don't collide with one another.
+     It is fine to lump them and merge them together before merging with the 'uncolliding' set.
+
+      innerColliding pipes collide with one another and must each be checkpointed before use in the general merge.
+     */
+    val safedInnerColliding = innerColliding.map(new Checkpoint(_))
+    val safedInnerUncolliding =
+      if (innerUncolliding.isEmpty) Nil
+      else List(new Checkpoint(mergeAvoidingNameClashes(innerUncolliding.head, innerUncolliding.tail)))
+
+    val reassembled = safedInnerColliding ::: safedInnerUncolliding ::: uncolliding
+    mergeAvoidingNameClashes(reassembled.head, reassembled.tail)
+  }
+
+  /**
+    * Cascading Merge does not support having multiple incoming pipes with the same name.
+    * Selectively rename pipes to avoid naming conflicts.
+    *
+    * @param head the first pipe to be merged (or checkpointed)
+    * @param tail a list of other pipes to be merged within the first
+    * @return a Merge of input pipes with any name clashes removed, or the input pipe if there was only one
+    */
+  private[scalding] def mergeAvoidingNameClashes(head: Pipe, tail: List[Pipe]): Pipe = tail match {
+    case Nil => head // avoid generating new Merge(pipes.head)
+    case _ =>
+      val (result, buf) = tail.foldLeft((List[Pipe](head), Set[String](head.getName))) {
+        case ((result, names), p) =>
+          if (names.contains(p.getName))
+            (assignName(p) :: result, names) // assignName is guaranteed unique and never assigned again
```

Comment:
we don't, since the result of assignName() is guaranteed unique (so it won't collide with pre-existing items from names) and will never be assigned again (so no pipe yet to be evaluated in the foldLeft will have a name that collides with the freshly-assigned pipe name — unless the developer's being intentionally naming a pipe _pipe_42, but that's a case of "doctor, my eye hurts when I drink coffee while leaving the spoon in the mug" that has always existed). 

I guess I need to pre-pend: "no need to add new name to 'names': " to the comment